### PR TITLE
gh-104773: PEP 594: Remove the imghdr module

### PR DESCRIPTION
**Pull request by vstinner**
* Remove the Lib/test/imghdrdata/ directory.
* Copy 5 pictures (gif, png, ppm, pgm, xbm) from removed Lib/test/imghdrdata/ to a new Lib/test/tkinterdata/ directory.

* Issue: gh-104773

----
:books: Documentation preview :books:: https://cpython-previews--104777.org.readthedocs.build/

### CODE REVIEW

**Comment by vstinner** [comment]
The Docs CI job fails with:

```
Could not import extension sphinx.builders.epub3 (exception: No module named 'imghdr')
```

Oh. It seems like the Sphinx EPUB builder requires the removed ``imghdr`` module.

cc @julienpalard @deronnax

**Comment by hugovk** [comment]
Sphinx used to depend on imghdr, but no longer does as of Sphinx 6.2.0:

* https://github.com/sphinx-doc/sphinx/issues/10440#issuecomment-1556180835
* https://github.com/sphinx-doc/sphinx/commit/a502e7523376e0344c1c9cc8a9d128143cd98b2d
* https://github.com/sphinx-doc/sphinx/releases/tag/v6.2.0

Therefore we'll need to start using Sphinx 6.2.0 for Python 3.13.

cc @AA-Turner @CAM-Gerlach.

**Comment by JulienPalard** [comment]
> Therefore we'll need to start using Sphinx 6.2.0 for Python 3.13.

@doko42 would it be managable from a Debian packaging point of view?

Do we have some contacts with other distribs?

I remember the last bump of `needs_sphinx` raised a lot of concerns.

**Comment by hugovk** [comment]
@encukou Is requiring Sphinx 6.2.0 for Python 3.13 okay for Red Hat?

File: Doc/whatsnew/3.12.rst [hunk 1]
```diff
@@ -809,7 +809,7 @@ Modules (see :pep:`594`):
 * :mod:`cgitb`
 * :mod:`chunk`
 * :mod:`crypt`
-* :mod:`imghdr`
+* ``imghdr``
```
Comment (hugovk) [inline Doc/whatsnew/3.12.rst]:
```suggestion
* :mod:`!imghdr`
```

File: Doc/whatsnew/3.13.rst [hunk 2]
```diff
@@ -102,6 +102,11 @@ Deprecated
 Removed
 =======
 
+* Remove the ``imghdr`` module, deprecated in Python 3.11: use the projects
+  `filetype <https://pypi.org/project/filetype/>`_,
+  `puremagic <https://pypi.org/project/puremagic/>`_,
+  or `python-magic <https://pypi.org/project/python-magic/>`_ instead.
+  (Contributed by Victor Stinner in :gh:`104773`.)
```
Comment (hugovk) [inline Doc/whatsnew/3.13.rst:252]: Reference the PEP.

File: Misc/NEWS.d/next/Library/2023-05-23-02-13-11.gh-issue-104773.JNiEjv.rst [hunk 3]
```diff
@@ -0,0 +1,4 @@
+Remove the ``imghdr`` module, deprecated in Python 3.11: use the projects
+`filetype <https://pypi.org/project/filetype/>`_, `puremagic
+<https://pypi.org/project/puremagic/>`_, or `python-magic
+<https://pypi.org/project/python-magic/>`_ instead. Patch by Victor Stinner.
```
Comment (hugovk) [inline Misc/NEWS.d/next/Library/2023-05-23-02-13-11.gh-issue-104773.JNiEjv.rst]: Reference the PEP.

**Review by hugovk — COMMENTED** [review]
Avoid new Sphinx warnings:
```diff
diff --git a/Doc/whatsnew/3.11.rst b/Doc/whatsnew/3.11.rst
index 7a479c6e56..9b33155c5a 100644
--- a/Doc/whatsnew/3.11.rst
+++ b/Doc/whatsnew/3.11.rst
@@ -1735,7 +1735,7 @@ Modules
   +---------------------+---------------------+---------------------+---------------------+---------------------+
   | :mod:`audioop`      | :mod:`crypt`        | :mod:`nis`          | :mod:`sndhdr`       | :mod:`uu`           |
   +---------------------+---------------------+---------------------+---------------------+---------------------+
-  | :mod:`cgi`          | :mod:`imghdr`       | :mod:`nntplib`      | :mod:`spwd`         | :mod:`xdrlib`       |
+  | :mod:`cgi`          | :mod:`!imghdr`      | :mod:`nntplib`      | :mod:`spwd`         | :mod:`xdrlib`       |
   +---------------------+---------------------+---------------------+---------------------+---------------------+
   | :mod:`cgitb`        | :mod:`mailcap`      | :mod:`ossaudiodev`  | :mod:`sunau`        |                     |
   +---------------------+---------------------+---------------------+---------------------+---------------------+
diff --git a/Doc/whatsnew/3.5.rst b/Doc/whatsnew/3.5.rst
index 14b6425cea..d13036f6fd 100644
--- a/Doc/whatsnew/3.5.rst
+++ b/Doc/whatsnew/3.5.rst
@@ -1252,7 +1252,7 @@ Oberkirch in :issue:`21800`.)
 imghdr
 ------
 
-The :func:`~imghdr.what` function now recognizes the
+The :func:`!imghdr.what` function now recognizes the
 `OpenEXR <https://www.openexr.com>`_ format
 (contributed by Martin Vignali and Claudiu Popa in :issue:`20295`),
 and the `WebP <https://en.wikipedia.org/wiki/WebP>`_ format
```

**Comment by encukou** [comment]
> Is requiring Sphinx 6.2.0 for Python 3.13 okay for Red Hat?

Yes. It's OK for Fedora. (And RHEL will follow Fedora after some time.)

**Comment by AA-Turner** [comment]
> Avoid new Sphinx warnings:
> 
> ```diff
> diff --git a/Doc/whatsnew/3.11.rst b/Doc/whatsnew/3.11.rst
> ...
> @@ -1252,7 +1252,7 @@ Oberkirch in :issue:`21800`.)
>  imghdr
>  ------
>  
> -The :func:`~imghdr.what` function now recognizes the
> +The :func:`!imghdr.what` function now recognizes the
> ...
> ```

We should likely prefer `` :func:`~!imghdr.what` `` to keep the rendered HTML showing only ``what``.

A

File: Doc/whatsnew/3.13.rst [hunk 4]
```diff
@@ -102,6 +102,11 @@ Deprecated
 Removed
 =======
 
+* Remove the ``imghdr`` module, deprecated in Python 3.11: use the projects
```
Comment (AA-Turner) [inline Doc/whatsnew/3.13.rst]:
```suggestion
* Remove the :mod:`!imghdr` module, deprecated in Python 3.11: use the projects
```

File: Misc/NEWS.d/next/Library/2023-05-23-02-13-11.gh-issue-104773.JNiEjv.rst [hunk 5]
```diff
@@ -0,0 +1,4 @@
+Remove the ``imghdr`` module, deprecated in Python 3.11: use the projects
```
Comment (AA-Turner) [inline Misc/NEWS.d/next/Library/2023-05-23-02-13-11.gh-issue-104773.JNiEjv.rst]:
```suggestion
Remove the :mod:`!imghdr` module, deprecated in Python 3.11: use the projects
```

**Comment by vstinner** [comment]
I created issue #104818 "Require Sphinx 6.2 to build Python 3.13 documentation". I propose to discuss updating Sphinx (or not) there.

**Comment by vstinner** [comment]
I rebased the PR on the main branch.

**Review by hugovk — APPROVED** [review]
Looks good, we can use Sphinx 6.2 for the "normal" docs build and for doctest (both use `requirements.txt`), and still support 3.2 in `Doc/requirements-oldest-sphinx.txt`.

**Comment by vstinner** [comment]
> Looks good, we can use Sphinx 6.2 for the "normal" docs build and for doctest (both use requirements.txt), and still support 3.2 in Doc/requirements-oldest-sphinx.txt.

Right. I only upgraded Sphinx in the GHA "Doctest" job. It seems like increasing the minimum Sphinx version to build the documentation is not **strictly required** to remove the imghdr module.

I prefer to decorelate the imghdr removal and increase the minimum Sphinx version to *build* the Python documentation. So far, still supporting Sphinx 3.2 makes it possible to build Python 3.13 documentation on Debian Stable: https://github.com/python/cpython/issues/104818#issuecomment-1564343407

**Pull request merged**